sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1585,3 +1586,110 @@ object ResolveUpCast extends Rule[LogicalPlan] {
+   * marked with a +, and invalid ones are marked with a x. The invalid ones are filtered using the
+   * Filter operator.
+   * window: 12m, slide: 5m, start: 0m :: window: 12m, slide: 5m, start: 2m
+   *     11:55 - 12:07 +                      11:52 - 12:04 x
+   *     12:00 - 12:12 +                      11:57 - 12:09 +
+   *     12:05 - 12:17 +                      12:02 - 12:14 +
+   *
+   * @param p The logical plan
+   * @return the logical plan that will generate the time windows using the Expand operator, with
+   *         the Filter operator for correctness and Project for usability.
+   */
+  private def generateWindows(p: LogicalPlan): LogicalPlan = {
+    val windowExpr = p.expressions.collect(getWindowExpr[TimeWindow](e => e)).head
+    // get all expressions we need to pass on for projections
+    val otherExpressions =
+      getMissingAttributes(p).filterNot(_.children.exists(_.isInstanceOf[TimeWindow]))
+
+    val projections = Seq.tabulate(windowExpr.maxNumOverlapping + 1) { i =>
+      // windowId <- ceil((timestamp - startTime) / slideDuration)
+      val division = Ceil(Divide(Subtract(Cast(windowExpr.timeColumn, LongType),
+        Literal(windowExpr.startTime)), Literal(windowExpr.slideDuration)))
+      // start <- (windowId + i - maxNumOverlapping) * slideDuration + startTime
+      // the 1000000 is necessary for properly casting a LongType to a TimestampType
+      val windowStart =
+        Multiply(
+          Add(
+            Multiply(
+              Add(division, Literal(i - windowExpr.maxNumOverlapping)),
+              Literal(windowExpr.slideDuration)),
+            Literal(windowExpr.startTime)),
+          Literal(1000000))
+      // windowEnd <- windowStart + windowDuration
+      val windowEnd =
+        Add(windowStart, Multiply(Literal(windowExpr.windowDuration), Literal(1000000)))
+      windowStart :: windowEnd :: windowExpr.originalTimeColumn :: Nil ++ otherExpressions
+    }
+    val timeCol = windowExpr.originalTimeColumn.references.toSeq
+    // timestamp >= window.start && timestamp < window.end
+    val filterExpr = And(GreaterThanOrEqual(windowExpr.timeColumn, windowExpr.windowStartCol),
+      LessThan(windowExpr.timeColumn, windowExpr.windowEndCol))
+    Project(windowExpr.output ++ Seq(windowExpr.outputColumn) ++ timeCol ++ otherExpressions,
+      Filter(filterExpr,
+        Expand(projections, windowExpr.output ++ timeCol ++ otherExpressions.map(_.toAttribute),
+          p.children.head)))
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case p: LogicalPlan if p.expressions.collect(getWindowExpr[TimeWindow](e => e)).nonEmpty &&
+      p.children.length == 1 =>
+      val windowed = generateWindows(p)
+      val rewritten = p transformExpressions getWindowExpr { windowExpr =>
+        windowExpr.validate() match {
+          case Some(e) => throw new AnalysisException(e)
```

Comment:
Whenever possible, we try to throw AnalysisExceptions from CheckAnalysis.  That way you can partially analyze a plan when debugging.  Maybe we could put the validation logic into `checkInputDataTypes` and then just leave TimeWindow expressions in the plan if they are not valid?